x-pack/plugins/fleet/server/services/agents/status.ts

```diff
@@ -37,88 +41,108 @@ export async function getAgentStatusById(
+    });
+  }
+
+  const query =
+    clauses.length > 0
+      ? {
+          bool: {
+            must: clauses,
+          },
+        }
+      : undefined;
+
+  const statuses: Record<AgentStatus, number> = {
+    online: 0,
+    error: 0,
+    inactive: 0,
+    offline: 0,
+    updating: 0,
+    unenrolled: 0,
+    degraded: 0,
+    enrolling: 0,
+    unenrolling: 0,
+  };
+
+  let response;
+
   try {
-    pitId = await openPointInTime(esClient);
+    response = await esClient.search<
+      null,
+      { status: AggregationsTermsAggregateBase<AggregationsStatusTermsBucketKeys> }
+    >({
+      index: AGENTS_INDEX,
+      size: 0,
+      query,
+      fields: Object.keys(runtimeFields),
+      runtime_mappings: runtimeFields,
+      aggregations: {
+        status: {
+          terms: {
+            field: 'status',
+            size: Object.keys(statuses).length,
+          },
+        },
+      },
+    });
   } catch (error) {
-    if (error.statusCode === 404) {
-      appContextService
-        .getLogger()
-        .debug('Index .fleet-agents does not exist yet, skipping point in time.');
-    } else {
-      throw error;
-    }
+    logger.error(`Error getting agent statuses: ${error}`);
+    throw error;
   }
-  const [all, allActive, online, error, offline, updating] = await pMap(
-    [
-      undefined, // All agents, including inactive
-      undefined, // All active agents
-      AgentStatusKueryHelper.buildKueryForOnlineAgents(),
-      AgentStatusKueryHelper.buildKueryForErrorAgents(),
-      AgentStatusKueryHelper.buildKueryForOfflineAgents(),
-      AgentStatusKueryHelper.buildKueryForUpdatingAgents(),
-    ],
-    (kuery, index) =>
-      getAgentsByKuery(esClient, {
-        showInactive: index === 0,
-        perPage: 0,
-        page: 1,
-        pitId,
-        kuery: joinKuerys(
-          ...[
-            kuery,
-            filterKuery,
-            agentPolicyId ? `${AGENTS_PREFIX}.policy_id:"${agentPolicyId}"` : undefined,
-          ]
-        ),
-      }),
-    {
-      concurrency: 1,
-    }
-  );
 
-  if (pitId) {
-    await closePointInTime(esClient, pitId);
+  const buckets = response?.aggregations?.status?.buckets || [];
+
+  if (!Array.isArray(buckets)) {
+    throw new Error('buckets is not an array');
   }
 
-  const result = {
-    total: allActive.total,
-    inactive: all.total - allActive.total,
-    online: online.total,
-    error: error.total,
-    offline: offline.total,
-    updating: updating.total,
-    other: all.total - online.total - error.total - offline.total,
+  buckets.forEach((bucket) => {
+    if (statuses[bucket.key] !== undefined) {
+      statuses[bucket.key] = bucket.doc_count;
+    }
+  });
+
+  const combinedStatuses = {
+    online: statuses.online,
+    error: statuses.error + statuses.degraded,
+    inactive: statuses.inactive,
+    offline: statuses.offline,
+    updating: statuses.updating + statuses.enrolling + statuses.unenrolling,
+    unenrolled: statuses.unenrolled,
```

Comment:
can any of these values be `undefined`? maybe we should fall back to 0

Reply:
looks like the code on top check for `undefined` no?

Reply:
As I understood, the code builds the `statuses` map from the dynamic buckets coming from the agg, and later tries to use specific values like `unenrolled`, but I think there is no guarantee that there is at least one match on that status in buckets.

Reply:
On line 77 I pre-create a map with all of the statuses defaulted to 0 meaning there isn't a risk of undefined here

Reply:
That's good then, sorry I didn't notice that.